discovery_engine/lib/src/domain/event_handler.dart

```diff
@@ -0,0 +1,217 @@
+    show DocumentRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/assets/assets.dart'
+    show createDataProvider, createManifestReader;
+import 'package:xayn_discovery_engine/src/infrastructure/assets/http_asset_fetcher.dart'
+    show HttpAssetFetcher;
+import 'package:xayn_discovery_engine/src/infrastructure/box_name.dart'
+    show activeDocumentDataBox, changedDocumentIdBox, documentBox;
+import 'package:xayn_discovery_engine/src/infrastructure/repository/hive_active_document_repo.dart'
+    show HiveActiveDocumentDataRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/repository/hive_changed_document_repo.dart'
+    show HiveChangedDocumentRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/repository/hive_document_repo.dart'
+    show HiveDocumentRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/type_adapters/hive_duration_adapter.dart'
+    show DurationAdapter;
+import 'package:xayn_discovery_engine/src/infrastructure/type_adapters/hive_unique_id_adapter.dart'
+    show DocumentIdAdapter, StackIdAdapter;
+import 'package:xayn_discovery_engine/src/infrastructure/type_adapters/hive_uri_adapter.dart'
+    show UriAdapter;
+import 'package:xayn_discovery_engine/src/logger.dart' show logger;
+
+const kEnginePath = 'engine';
+const kDatabasePath = 'database';
+
+class EventHandler {
+  Future<Engine>? _engineFuture;
+  late final DocumentRepository _documentRepository;
+  late final ActiveDocumentDataRepository _activeDataRepository;
+  late final ChangedDocumentRepository _changedDocumentRepository;
+  late final DocumentManager _documentManager;
+  late final FeedManager _feedManager;
+
+  EngineExceptionRaised _handleInitError(Object e) {
+    logger.e(e);
+    _engineFuture = null;
+    final reason = e is AssetFetcherException
+        ? EngineExceptionReason.failedToGetAssets
+        : EngineExceptionReason.genericError;
+    return EngineExceptionRaised(reason);
+  }
+
+  /// Decides what to do with incoming [ClientEvent] by passing it
+  /// to a dedicated manager and returns the appropriate response in the in form
+  /// of a proper [EngineEvent].
+  ///
+  /// This handler is invoked by a [ClientEvent]s stream listener, so it is
+  /// called each time there is a new event on the stream, without waiting for
+  /// previous events to finish processing.
+  Future<EngineEvent> handleMessage(ClientEvent clientEvent) async {
+    if (clientEvent is Init) {
+      try {
+        _engineFuture = _initEngine(clientEvent.configuration);
+        await _engineFuture;
+        return const EngineEvent.clientEventSucceeded();
+      } catch (e) {
+        return _handleInitError(e);
+      }
+    }
+
+    if (_engineFuture == null) {
+      return const EngineEvent.engineExceptionRaised(
+        EngineExceptionReason.engineNotReady,
+      );
+    }
+
+    try {
+      await _engineFuture;
+    } catch (e) {
+      return _handleInitError(e);
```

Comment:
```suggestion
      return EngineExceptionReason.engineNotReady
```

I just realized there is no reason to handle the same error multiple times.